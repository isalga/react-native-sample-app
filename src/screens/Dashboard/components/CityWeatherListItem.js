// @flow
import type { City, Weather } from '../../../entities/types'
import React from 'react'
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native'
import { Icon } from 'react-native-elements'
import { connect } from 'react-redux'
import { loadStart as loadWeatherStart } from '../../../entities/weather/actions'
import { setSelectedId, removeCity } from '../../../entities/city/actions'

type Props = {
  city: ?City,
  weather: ?Weather,
  selectCity: (id: string) => void,
  removeCity: (id: string) => void,
  loadWeather: (city: City) => void,
  navigate: (screen: string) => void
}

export class CityWeatherListItem extends React.Component<Props> {

  static defaultProps = {
    loadWeather: () => {},
    selectCity: () => {},
    removeCity: () => {},
    navigate: () => {}
  }

  componentDidMount() {
    if (!this.props.weather && this.props.city)
      this.props.loadWeather(this.props.city)
  }

  onPressItem = () => {
    if (this.props.city) {
      this.props.selectCity(this.props.city.id)
      this.props.navigate('CityWeather')
    }
  }

  onRemoveItem = () => {
    if (this.props.city) {
      this.props.removeCity(this.props.city.id)
    }
  }

  render() {
    const { city, weather } = this.props
    if (!city || typeof city != 'object')
      return null

    return (
      <View style={styles.listItem}>
        <TouchableOpacity
          onPress={this.onPressItem}>
          <Text>
            { city.displayName } - { weather ? weather.temperature : '...loading' }
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={this.onRemoveItem}>
          <Icon name='delete'/>
        </TouchableOpacity>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  listItem: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    padding: 10,
    height: 44
  },
})

export default connect(
  (state, { city }) => ({
    weather: state.weather[city.id]
  }),
  (dispatch) => ({
    selectCity: (id) => {
      dispatch(setSelectedId(id))
    },
    removeCity: (id) => {
      dispatch(removeCity(id))
    },
    loadWeather: (city) => {
      dispatch(loadWeatherStart(city))
    }
  })
)(CityWeatherListItem)
