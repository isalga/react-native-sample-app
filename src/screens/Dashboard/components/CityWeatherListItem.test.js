import React from 'react'
import ReactDOM from 'react-dom'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { CityWeatherListItem } from './CityWeatherListItem'

Enzyme.configure({ adapter: new Adapter() })

const city = {
  id: '1',
  displayName: 'Abu Dhabi',
  name: 'Abu_Dhabi',
  country: 'united_arab_Emirates'
}

const weather = {}

describe('CityWeatherListItem component', () => {

  it('should call props.loadWeather', () => {
    const loadWeather = jest.fn()
    const wrapper = Enzyme.shallow(
      <CityWeatherListItem
        city={city}
        loadWeather={loadWeather}/>)
    expect(loadWeather.mock.calls.length).toEqual(1)
  })

  it('should call props.selectCity and props.navigate onPressItem', () => {
    const selectCity = jest.fn()
    const navigate = jest.fn()
    const wrapper = Enzyme.shallow(
      <CityWeatherListItem
        city={city}
        selectCity={selectCity}
        navigate={navigate}/>)

    wrapper.instance().onPressItem()

    expect(selectCity.mock.calls.length).toEqual(1)
    expect(navigate.mock.calls.length).toEqual(1)
  })

  it('should call props.removeCity onRemoveItem', () => {
    const removeCity = jest.fn()
    const wrapper = Enzyme.shallow(
      <CityWeatherListItem
        city={city}
        removeCity={removeCity}/>)

    wrapper.instance().onRemoveItem()

    expect(removeCity.mock.calls.length).toEqual(1)
  })

  it('should render correctly', () => {
    const wrapper = Enzyme.shallow(
      <CityWeatherListItem
        city={city}/>)
    expect(wrapper).toMatchSnapshot()
  })
})
