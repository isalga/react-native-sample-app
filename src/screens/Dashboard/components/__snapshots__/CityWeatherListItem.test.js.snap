// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityWeatherListItem component should render correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "height": 44,
      "justifyContent": "space-around",
      "padding": 10,
    }
  }
>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Abu Dhabi
       - 
      ...loading
    </Text>
  </TouchableOpacity>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
  >
    <Icon
      color="black"
      name="delete"
      raised={false}
      reverse={false}
      reverseColor="white"
      size={24}
      underlayColor="white"
    />
  </TouchableOpacity>
</View>
`;
